pandas/tools/merge.py

```diff
@@ -110,46 +215,232 @@ def ordered_merge(left, right, on=None, left_by=None, right_by=None,
-                           suffixes=suffixes, fill_method=fill_method)
+                           suffixes=suffixes, fill_method=fill_method,
+                           how=how)
         return op.get_result()
 
     if left_by is not None and right_by is not None:
         raise ValueError('Can only group either left or right frames')
     elif left_by is not None:
-        if not isinstance(left_by, (list, tuple)):
-            left_by = [left_by]
-        pieces = []
-        for key, xpiece in left.groupby(left_by):
-            merged = _merger(xpiece, right)
-            for k in left_by:
-                # May have passed ndarray
-                try:
-                    if k in merged:
-                        merged[k] = key
-                except:
-                    pass
-            pieces.append(merged)
-        return concat(pieces, ignore_index=True)
+        result, _ = _groupby_and_merge(left_by, on, left, right,
+                                       lambda x, y: _merger(x, y),
+                                       check_duplicates=False)
     elif right_by is not None:
-        if not isinstance(right_by, (list, tuple)):
-            right_by = [right_by]
-        pieces = []
-        for key, ypiece in right.groupby(right_by):
-            merged = _merger(left, ypiece)
-            for k in right_by:
-                try:
-                    if k in merged:
-                        merged[k] = key
-                except:
-                    pass
-            pieces.append(merged)
-        return concat(pieces, ignore_index=True)
+        result, _ = _groupby_and_merge(right_by, on, right, left,
+                                       lambda x, y: _merger(y, x),
+                                       check_duplicates=False)
     else:
-        return _merger(left, right)
+        result = _merger(left, right)
+    return result
+
+
+def merge_asof(left, right, on=None,
+               left_on=None, right_on=None,
+               by=None,
+               suffixes=('_x', '_y'),
+               tolerance=None,
+               allow_exact_matches=True,
+               check_duplicates=True):
+    """Perform an asof merge. This is similar to a left-join except that we
+    match on nearest key rather than equal keys.
+
+    For each row in the left DataFrame, we select the last row in the right
+    DataFrame whose 'on' key is less than or equal to the left's key. Both
+    DataFrames must be sorted by the key.
+
+    Optionally perform group-wise merge. This searches for the nearest match
+    on the 'on' key within the same group according to 'by'.
+
+    .. versionadded 0.18.2
+
+    Parameters
+    ----------
+    left : DataFrame
+    right : DataFrame
+    on : label or list
+        Field names to join on. Must be found in both DataFrames.
+        The data MUST be ordered. Furthermore this must be numeric index,
+        typically a datetimelike or integer-index. On or left_on/right_on
```

Comment:
I find the use of the word 'index' a bit confusing here. As it are not 'index' in the pandas meaning (maybe rather use 'key'?)